v1/lists.js

```diff
@@ -0,0 +1,151 @@
+            }
+            // Make sure nothing malicious can be done by splicing the continuation data
+            // into the API parameters.
+            const allowedKeys = ['continue', 'rlcontinue', 'rlecontinue'];
+            for (const key of allowedKeys) {
+                if (typeof continuation[key] !== 'object') {
+                    sanitizedContinuation[key] = continuation[key];
+                }
+            }
+        }
+        return sanitizedContinuation;
+    }
+
+    /**
+     * Convert an array of values into the format expected by the MediaWiki API.
+     * @param {!Array} list A list containing strings and numbers.
+     * @return {!String}
+     */
+    flattenMultivalue(list) {
+        return list.join('|');
+    }
+
+    /**
+     * Handle the /list/{id}/entries endpoint (get entries of a list).
+     * @param {!HyperSwitch} hyper
+     * @param {!Object} req The request object as provided by HyperSwitch.
+     * @return {!Promise<Object>} A response promise.
+     */
+    getListEntries(hyper, req) {
+        return hyper.post({
+            uri: new URI([req.params.domain, 'sys', 'action', 'rawquery']),
+            body: {
+                action: 'query',
+                list: 'readinglistentries',
+                rlelists: req.params.id,
+                rlelimit: 'max',
+                continue: this.unflattenContinuation(req.query.next).continue,
+                rlecontinue: this.unflattenContinuation(req.query.next).rlecontinue,
+            },
+        })
+        .then((res) => {
+            const entries = res.body.query.readinglistentries;
+            const next = this.flattenContinuation(res.body.continue);
+
+            return this.hydrateSummaries(entries, hyper, req)
+            .then(entries => ({
+                status: 200,
+                headers: {
+                    'content-type': 'application/json; charset=utf-8;'
+                        + 'profile="https://www.mediawiki.org/wiki/Specs/Lists/0.1"',
+                    'cache-control': 'no-cache',
+                },
+                body: {
+                    entries,
+                    next,
+                }
+            }));
+        });
+    }
+
+    /**
+     * Add data from the summary endpoint to an array of list entries.
+     * @param {!Array} entries An array of objects with two properties:
+     *   - project: the domain of the wiki
+     *   - title: the title of the page
+     * @param {!HyperSwitch} hyper
+     * @param {!Object} req The request object as provided by HyperSwitch.
+     * @return {!Promise<Array>} The objects, enriched with summaries.
+     */
+    hydrateSummaries(entries, hyper, req) {
+        return mwUtil.getSiteMatrix(hyper, req).then((sitematrix) => {
+            return P.map(entries, (entry) => {
+                const site = sitematrix[entry.project];
+                if (site && !site.private) {
+                    const uri = `https://${entry.project}/api/rest_v1/page/summary/`
```

Comment:
So this is an interesting one here: if you go for the `/rest_v1` URI the request will leave RESTBase and hit Varnish. In case it's a cache miss in varnish it will come back to RESTBase and go to Cassandra for the summary. When doing these hydrations before for other things, going directly to Cassandra here (using internal `/sys` summary URI) was slightly faster. However that was before the PagePreviews feature was introduced, so most likely now the hit ratio in Varnish will be better and going to Varnish will be slightly faster.

Having said that, I still think we need to go to the internal URI for summary here. The difference in latency will be negligible but going to Varnish will mess up all the request rate statistics here. 

Here's an example how to go directly to RESTBase: https://github.com/wikimedia/restbase/blob/master/v1/related.js#L38-L39

Reply:
@Pchelolo Couldn't we still go to varnish but give it a special X-Analytics header value to indicate that this is an internal request?

Reply:
@berndsi why bother? there's no real benefit in doing that..

Reply:
hm @tgr interesting question and a very valid concern.. Ok, I think I'd be fine with going to Varnish here. 

One thing is that I'd avoid hard-coding `/api/rest_v1` - in the `mwUtils.getSiteInfo` we have a property `baseURITemplate` that's configurable - better to use that to construct the URIs so that if a wiki has no `api/rest_v1` rewrite RESTBase would still be usable.

Reply:
@Pchelolo My comment was just prompted by your earlier comment:

> so most likely now the hit ratio in Varnish will be better and going to Varnish will be slightly faster.

Not saying that we have to do it, obviously, just wanted to note that we could if the benefits outweigh the cons.